### ISSUE TITLE
UI Simplification: Publish Story Modal

### ISSUE THREAD

**Issue by miina** [swissspidy/web-stories-wp#10115]
## Feature Description
See [Figma](https://www.figma.com/file/ZrO763yvMWJjYeAPsRnNfW/_Editor---Sprint-Feature-Level-Designs_2022?node-id=57%3A14792)

- [x] When clicking on "Publish", display a modal with various relevant inputs

Breaking this up into chunks to make review cleaner, it might actually sustain more than 5 points depending on implementation and shared code. 

- [x] New modal component for publish story (new header and different spacing patterns so let's not reuse the standard dialog in the design system) 
- [x] Publish button to be disabled if fields in second column are empty (new) 
- [x] Unit tests 
- [x] Karma tests
- [x] New experiment 
- [x] Make inputs in second column (story title and description, password protected requires a password) mandatory 
- [x] Can we reuse the inspector panel contents? How is the best way to share this across this implementation as well as the tab structure? 
- [x] Pull in functionality for saving and updating story title 
- [x] Pull in functionality for saving and updating story description  _separate_ from document panel (where it lives now)
- [x] Pull in functionality for saving and updating status and visibility of story _separate_ from document panel (where it lives now)
- [x] Close story details modal and open checklist when checklist button is clicked 
- [x] Cover preview is brand new functionality - we're not showing a cover preview at all right now. Can we pull this in from AMP, are we doing it ourselves? + Update story cover functionality (when user clicks pencil icon does the story details modal close and open media picker? Does Story Details modal reopen after new image is selected or cancelled?)
- [x] Replace existing modal that opens when first publish occurs with this new one. No more 'are you sure you want to publish' with checklist flags, this new one will suffice.
- [x] e2e tests for functionality at different roles

## Additional Context
<img width="500" alt="Screenshot 2022-01-06 at 13 59 25" src="https://user-images.githubusercontent.com/3294597/148379890-feac3eea-f518-4133-a981-10022a67a60f.png">

**Comment by miina** [comment]
To be defined:
- Will this modal be always opened or only if there are checklist issues?
- Details of the fields that are displayed in the modal. Will only fields with issues be displayed? Or the full Document panel?
- What happens when clicking on "Checklist", will the modal close and checklist open? Or will this replace the checklist?
- How does this influence any currently existing confirmation dialogues we have? For example when changing a Draft Story status to Private, this appears:
<img width="907" alt="Screenshot 2022-01-06 at 13 55 50" src="https://user-images.githubusercontent.com/3294597/148379574-ccff001e-03d4-4105-bde5-44310b27355c.png">

Also, when publishing a story, this appears:
<img width="414" alt="Screenshot 2022-01-06 at 13 58 27" src="https://user-images.githubusercontent.com/3294597/148379791-7c5f47b4-d9c8-4011-8613-4ce344e90e73.png">

Not sure if there are any others.

**Comment by miina** [comment]
cc @swissspidy Tagging you here as well just in case since you've been working on different confirmation dialogues related to publishing. Perhaps you have some comments/concerns.

**Comment by agingoldseco** [comment]
Some answers:

> Will this modal be always opened or only if there are checklist issues?

It will always be opened for unpublished stories. 

> Details of the fields that are displayed in the modal. Will only fields with issues be displayed? Or the full Document panel?

All fields will be displayed for consistency in this modal and to allow the creator to change anything they may have put in before. 

> What happens when clicking on "Checklist", will the modal close and checklist open? Or will this replace the checklist?

Unsure right now—I think we should wait on having a checklist button at all until we figure out how to build it into the publish modal, which is somewhat out of scope. Or, until then, we have the checklist button close the modal and open the checklist, perhaps just having people open the publish modal again after. Will confirm this ASAP. This modal definitely will not replace the checklist. 

> How does this influence any currently existing confirmation dialogues we have?

I think this wouldn't change either of those dialogue boxes—private publishing likely doesn't need all these steps first. The second dialogue box you shared can appear after hitting the publish button in the modal. 

@aaskedall Is this all correct?

**Comment by miina** [comment]
cc @BrittanyIRL For visibility

**Comment by swissspidy** [comment]
I really hope that with this modal creators won't forget to set a story poster anymore

**Comment by BrittanyIRL** [comment]
@miina @barklund do you all have any specific visions for how functionality is getting shared for this new modal with existing components? IE) story description exists in document panel right now - but it's wrapped in a panel and has highlighting. I can pull out the functionality to share it, isolate the children, or i can make a second implementation of it to avoid conflating it? 

This is i think a small example, a larger one would be all the other inspector panel tabs getting added to the modal. It'd be ideal to share that info but the way the panels get passed around is a bit involved. 

I'd love to get your thoughts since you're both working more deeply in this redesign before I go in making my own choices. 

Here's my initial sketch of updating story title and excerpt: https://github.com/GoogleForCreators/web-stories-wp/pull/10403

**Comment by miina** [comment]
>  I can pull out the functionality to share it, isolate the children, or i can make a second implementation of it to avoid conflating it?

On Design Panel: Since generally the Design panel inside the Modal should match the Design panel of the editor then would be ideal if we could use the same components so that we wouldn't need to duplicate the changes every time.

On document panel items:
I guess it would be ideal if those could also share the functionality or utils or constants etc. and just differ in rendering, however, not sure if there are any restrictions doing that. For example, the slug has min-max limits and specific logic how to handle updates / cleaning the slug etc., it would be great if changing in one place, it would also change in another place, e.g. by adding `useSlug` or something similar in this specific case. Seems more maintainable compared to a full second implementation. Thoughts?

**Comment by BrittanyIRL** [comment]
> On Design Panel: Since generally the Design panel inside the Modal should match the Design panel of the editor then would be ideal if we could use the same components so that we wouldn't need to duplicate the changes every time.

I'm not actually sure what parts of the design panel are getting shared here? https://github.com/GoogleForCreators/web-stories-wp/tree/main/packages/story-editor/src/components/panels/design unless you meant another? 

> On document panel items: I guess it would be ideal if those could also share the functionality or utils or constants etc. and just differ in rendering, however, not sure if there are any restrictions doing that. For example, the slug has min-max limits and specific logic how to handle updates / cleaning the slug etc., it would be great if changing in one place, it would also change in another place, e.g. by adding `useSlug` or something similar in this specific case. Seems more maintainable compared to a full second implementation. Thoughts?

Yeah, sharing some utils as hooks is kinda where I'm at too after thinking a bit about this over the weekend. I wish I could just  import the whole document panel in and call it a day, but the options shown are slightly different (example a) the story excerpt isn't a panel, it's in the middle section). And then there's the [whole highlight functionality](https://github.com/GoogleForCreators/web-stories-wp/blob/main/packages/story-editor/src/components/panels/document/excerpt/excerpt.js#L59-L65) that this modal doesn't need to be aware of at all. I am probably worrying ahead of myself that by trying to share more logic it's going to clutter up both implementations. 

It's good to know that you're kinda on the same place though, creating utils that both can use as we see fit. I'll try to do this piecemeal so we can have conversation as we go. I'll update my draft for a shared hook with the excerpt and story title tomorrow to continue this - I'm taking today off.

**Comment by miina** [comment]
> I'm not actually sure what parts of the design panel are getting shared here?

Sorry, I meant the Document panel parts that will displayed in the Modal exactly as they are in the right-side panel as well! 🤦  My fingers and thoughts were not connected when typing "Design Panel".

But looks like you already pointed out why it might not be so straightforward so just sharing the parts that are possible to share makes sense! We could check later when there's more to compare if there are more parts to combine, might be hard to see in advance what will be the exact differences.

**Comment by BrittanyIRL** [comment]
Hey @aaskedall - since Amy's out this week we're probably not having a design sync - these are my current outstanding questions on this new feature. I think handling them async will be fine. 

When I’m in the editor and click “private” on story visibility it’ll automatically go to publish my story. Looping this ‘visibility’ section into the editor - what does that mean for the ‘private’ radio button in the modal? If I click “private” does the story just publish? Or does this status not persist until publish is manually selected in the modal? If the latter, it means that visibility will not be in sync with the editor panel’s visibility. 


Cover image and publisher logo both have dialogs of their own to select new media. Is this a dialog opening on top of this one? Does this one close for the media editor to take priority and then the user goes back to “publish” to get to the publish modal again? 

Cover preview has an edit media button (so same question about modals on modals) but it’s also the same exact thing as updating the cover image - do we want both of these? It’s duplicative. 

Cover preview is brand new! This is actually its own line item in the roadmap (currently titled “Publishing Discover Preview”) and it’s been kicked down the road for like 8 months or more because it might be obsolete with upcoming STAMP preview mode feature? This modal is ALOT as is, thoughts on v1 of it not having the cover preview at all? Just make the middle panel a bit wider to take the space initially.

**Comment by andrew-askedall** [comment]
-When I’m in the editor and click “private”...
Imho I think it's odd that private immediately publishes it. For public or password protected a user has to hit publish to make it "live" Can we have consistency withe the editor document panel if the radials stay in sync? We probably should treat it the same way as the other 2 radials where a user has to click publish manually. What are you thoughts?

-Cover image and publisher logo both have....
publisher logo has a drop down so I think that would fit over this modal. As far as the media selection model, it does appear to take up the whole screen so its not clear if the modal below closed or not. If a user backs out of the new modal it should return them to the publish modal. 

-Cover preview has a...
This was tough, I didn't want to redo the right side panel. I am worried users will look there and maybe miss the new preview. So that was my logic. I'm aware the button opens the same modal. I think it's more of like here is an abstracted version and here is the wysiwyg version. A user may also not open those other accordion menus. 

-I think it might be really beneficial to do a light weight version of this. It's basically a mirror of the title text, aligned bottom left and can't go more than 3 lines then it adds ellipses. I can spec it out better but I think it would help users with better cover composition. 
![image](https://user-images.githubusercontent.com/6991691/153022966-b8994ecb-8340-4bba-ade1-0cbcabc647c8.png)

**Comment by BrittanyIRL** [comment]
>>When I’m in the editor and click “private” on story visibility it’ll automatically go to publish my story.

> Imho I think it's odd that private immediately publishes it. For public or password protected a user has to hit publish to make it "live" Can we have consistency withe the editor document panel if the radials stay in sync? We probably should treat it the same way as the other 2 radials where a user has to click publish manually. What are you thoughts?

@miina @swissspidy I'm curious your takes here, if this immediate publish is important for some reason in the background? I agree with Andrew that it feels weird and I would love for it to be consistent.

**Comment by swissspidy** [comment]
> Imho I think it's odd that private immediately publishes it. For public or password protected a user has to hit publish to make it "live" Can we have consistency withe the editor document panel if the radials stay in sync? We probably should treat it the same way as the other 2 radials where a user has to click publish manually. What are you thoughts?

I need to give you some background on this.

This is a WordPress specific quirk. In essence, WordPress does not support having private draft posts. Private means immediately publishing as private. We cannot change this.

**Comment by BrittanyIRL** [comment]
> This is a WordPress specific quirk. In essence, WordPress does not support having private draft posts. Private means immediately publishing as private. We cannot change this.

Ah drag, that's what i was hoping wasn't the case. 

@andrew-askedall how do you want to incorporate this in this modal. We'd talked about making the title and description mandatory in the modal but if the user first updates publish to private then it'll override the disabled publish. Lemme know if you want to keep our sync to discuss!!

**Comment by BrittanyIRL** [comment]
@felipebochehin87 QA for this is here: https://github.com/GoogleForCreators/web-stories-wp/pull/10622

**Comment by felipebochehin87** [comment]
Tested against https://stories-qa-wordpress-amp.pantheonsite.io/wp-admin/, using PR #10622.

1. New modal is displayed as expected
2. Works fine in both RTL and LTR modes
3. Cover Preview looks good
4. Checklist works as expected
5. Publish button works as expected
6. Center pane looks good
7. Document pane looks good
8. Usability (keyboard and mouse) working as expected

**Comment by swissspidy** [comment]
@BrookeGraham with #10622 merged, can this be closed or is there still some work left to be done?
